### ISSUE TITLE
Scehma 2.0.0 Rework the O3DE object json schema and convert all canonical json

### ISSUE THREAD

**Issue by byrcolin** [o3de/o3de#17505]
The current schema 1.0.0 is insufficient for complex licensing, lacks consistency and schema. Schema 2.0.0 will introduce a new schema and provide consistency across all o3de objects.

**Comment by byrcolin** [comment]
Work is progressing. I have a prelim schema for 2.0.0. I'm going to upload that first to the website so validation can work without local files.

**Comment by byrcolin** [comment]
I have gotten a lot of work done recently. It is quite a large change, I am currently in the process of finishing the schema itself for RFC. Since this is very large I am going to use this issue as the master issue and create many smaller child issues that this issue links to. I have had a request to break out the versioning changes as its own issue, so that will be the first child issue I work on. Should be ready sometime soon, I am aiming for this week, but this is a challenging week in the US so we will see.

**Comment by kberg0** [comment]
Quick feedback you have some silly typos in there, differeniate and unmatinable and probably others. The occasional digression into first person narratives mixed with hyperbolic language could be removed entirely given this is a public document.
‘Not all bad by any means’ delete, 
‘But it also introduced what I and many others consider to be a completely unnacceptable notion of "child _data" entries.’ could just be ‘It introduced a problematic notion of “child_data” entries’ and then continue on with enumerating the issues.

Beyond that, you have a specific goal that inspired a lot of this effort, and I suspect it ballooned into more of a cleanup / tech debt reduction effort. The problem with tech debt reduction is that it can be quite disruptive, and I suspect that might be the case here. This is unfortunate because that disruption is now potentially tied to the success or failure of your original goal (restricted platform support if I’m not mistaken)

I would consider breaking this into two proposals, those that are the minimal critical components required for your efforts, and then the larger object schema 2.0 work which will face potential complications that will impact and disrupt current o3de customers, of which there are more than a few now. This means even if schema 2.0 ends up requiring a complex rollout process or even turns out to be infeasible, your core efforts to unblock restricted platforms is not blocked.

**Comment by KylerSF** [comment]
After talking in Discord on ways to stop gems that cannot work together, but can work in place of each other, (PhysX4 and PhysX5) Nick found the use of **provided_unique_service** in the PhysX gem.json files, but it does not seem to be working as expected. Though this getting fixed would allow two gems like PhysX4/5 to both have **"provided_unique_service": "Physics"** and ensure during the cmake phase of building a project that two gems that both provide a unique service do not share that same service. We should make sure Schema 2.0.0 either transfers this tag directly, or takes in a similar tag with the same functionality.

**Comment by ericliusunniy** [comment]
I've written up an alternative object model and filed it as #19989. Since this issue is the master RFC for Schema 2.0.0, let me be explicit about how the two relate, because calling it an "alternative" overstates the disagreement.

## What I inherit from this RFC unchanged

A formal, validatable schema referenced by `$schema` / `$schemaVersion`; dropping the child `_data` copying; grouping the root-level field sprawl; `licenses[]` as an SPDX-shaped array; `uri` over `url`; every object carrying its own license so it stays legible if it is ever separated from its parent. #19989 keeps all of it, and the reasoning in this issue for each is better than anything I could restate.

The item on separating `external_subdirectories` from gems deserves underlining, because it has the best cost-to-benefit ratio of anything here and it doesn't depend on the rest being accepted. #19989 §4 splits the concept in two: `gem_roots[]` are roots scanned for objects, `build_roots[]` are directories CMake recurses into that are deliberately *not* O3DE objects, and it's a generate-time error (`E2002`) for the two to be ancestors of each other or for a `gem.json` to appear inside `build_roots[]`. Same disambiguation this RFC asks for, with the "scan everything and then discard whatever lacks a gem.json" step deleted.

## Where I diverge

Three things, which is narrower than the length of my document suggests.

**1. The repository-split premise.** Item 5 says "this will soon no longer happen as the direction of the repo is to move these objects out of the core repo." #19989 takes the opposite premise: the monorepo stays, and the Schema's job is to describe the engine tree plus project tree precisely. My concern is concrete rather than ideological — today a change spanning AzCore, Atom, and PhysX is one PR, one CI run, and one revertible commit; across ~120 repositories it becomes three PRs, three version bumps, a lockfile update, and a manually ordered merge, where any lagging step leaves the mainline in an intermediate state. The atomic commit is currently the only cross-component regression barrier the project has. Repository governance seems worth discussing on its own, but I don't think it should be an implicit prerequisite of a Schema upgrade.

**2. `releases[]`, `download`, and `remote` as core Schema fields.** I think the need behind them is real — users must be able to find and obtain a Gem — but it reads to me as a launcher and Marketplace need rather than a build contract need. In #19989 the acquisition action lands source in the tree and writes one `gem_roots[]` entry, and everything after that is deterministic offline resolution. That way the Marketplace can evolve, or fail, without affecting any checked-out workspace. @nick-l-o3de's note that raw githubusercontent is firewalled in some countries is one more argument for keeping the build path from depending on it at all.

**3. Version-qualified dependencies resolved by selection.** Engine Gems ship in the same commit as the engine and are tested by the same CI run, so their compatibility is already guaranteed for free by the commit; external Gems genuinely need an explicit version contract. Flattening the two means giving up the consistency the first class already has and buying it back with ranges and solving. #19989 asserts versions and fails loudly instead of selecting among candidates — one name resolves to exactly one instance per closure, no solver.

## Picking up specific feedback already in this thread

**Naming convention (item 4).** @nick-l-o3de's point about tying names to real domains is exactly why #19989 does not mandate reverse-DNS. Identity is a stable UUID, `name` needs only closure-local uniqueness, and `namespace` is optional for organizations that want it. Because references match on id, a rename can't break existing references, and `name_aliases[]` covers the commercial-fork case.

**Licenses.** Also agreeing with Nick here: `is_deprecated` and `is_osi_approved` are facts that can change after you ship, so baking them into each object does work against self-description. #19989 keeps `licenses[]` SPDX-shaped but sources those mutable attributes from a versioned SPDX catalog, with the catalog version and digest entering the invalidation key.

**"NOT iterating into a subfolder of a gem using cmake if the gem in question is not part of the project being compiled."** This is answered directly. CMake consumes a resolved closure (`resolved-gems.json`) and wires only the directories in it, so a Gem that was discovered but not enabled is never recursed into. Discovery and enablement are separate steps by construction (§4, §5, §10).

**Child-of-child ambiguity.** Fixed discovery roots make this a non-question: Gems are found under `<engine>/Gems` and `<project>/Gems`, and there is no third tier to reason about.

**@KylerSF on `provided_unique_service`.** My draft didn't answer this — §5 declined to add any conflicts mechanism and waved at "feature configuration or module gating," which doesn't cover PhysX4/PhysX5 at all. You're right that 2.0 has to carry the capability across, so I've specified it rather than leave it implicit:

- `unique_services[]` on `gem.json` declares the subsystem slots a Gem occupies, and 1.x `provided_unique_service` migrates to it mechanically with the value transferred verbatim, so the existing PhysX exclusion survives the migration unchanged (§7.4, §12.2).
- The rule is one static check per build combination, evaluated after the active closure is fixed: more than one active provider of a service is `E2042` (§5.7, §13). The diagnostic has to name every provider with its scope, path, and *how it entered the closure*, because the usual cause is that the second one arrived through a transitive dependency rather than a direct reference.
- Both providers stay discoverable and may sit in the same engine tree — the check is scoped to the active closure of one combination, so keeping PhysX4 and PhysX5 both checked out and switching between them stays a supported workflow, and gating may legally activate different providers in different combinations.
- `E2042` is terminal. The resolver **must not** drop, deactivate, or substitute a provider to satisfy the rule, and there is deliberately no way to *depend on* a service — "require any Gem providing Physics" would reintroduce candidate enumeration, and choosing a candidate is solving. Dependencies stay name-based, which is what keeps this a check rather than a solve.
- §18.1 scenario 41 covers it end to end: the passing single-provider case, the direct double-enable, the transitive double-enable, the per-configuration gating case, duplicate declaration inside one Gem, the Overlay prohibition, case/normalization collisions, and two providers in the engine's own `default_enabled_gems[]` failing in engine-standalone mode with no project present.

Worth noting this is a case where the closure model is *easier* than a solver would be — no back-tracking, no candidate sets, just a set-cardinality check over an already-resolved list. If the check is right in shape, `provided_unique_service` also stops being PhysX-specific tribal knowledge and becomes something a validator can enforce for any subsystem.

## One process request

@kberg0's advice above — split this into the minimal critical components versus the larger schema work, so that a complex or infeasible rollout of the latter doesn't block the former — is the part of this thread I agree with most, and it applies to my document at least as much as to this one. §14.0.1 does that split: **RFC-A** (objects and resolution) is self-contained, touches neither the build nor the runtime, and delivers multi-engine registration, `external_subdirectories` convergence, and registration-free project Gems on its own; **RFC-B** (modules and loading) and **RFC-C** (overlays and override auditing) come after and may only add fields, never redefine RFC-A's.

@byrcolin, since you mentioned using this as the master issue with smaller child issues linked from it, would it make sense to treat RFC-A as one of those children so the parts that clearly overlap can be settled without waiting on the parts that don't? Also, this issue is labeled `sig/platform` whereas I guessed SIG-Core in #19989 — happy to be corrected on ownership, and I'll bring it to whichever SIG triage is right.

Full document (~184k characters, so it lives in a Gist rather than an issue body): https://gist.github.com/ericliusunniy/5ed1d4e842071584aeda9d552d4bea24 — the summary is in #19989.
